org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/FileEventHandler.java

```diff
@@ -129,49 +149,129 @@ private static WorkspaceEdit computePackageRenameEdit(FileRenameEvent[] renameEv
-						String newPackageName = String.join(".", newLocation.makeRelativeTo(sourceLocation).segments());
-						IPackageFragment oldPackageFragment = packageRoot.getPackageFragment(oldPackageName);
-						if (oldPackageFragment != null && !oldPackageFragment.isDefaultPackage() && oldPackageFragment.getResource() != null) {
-							oldPackageFragment.getResource().refreshLocal(IResource.DEPTH_INFINITE, null);
-							if (oldPackageFragment.exists()) {
-								ResourcesPlugin.getWorkspace().run((pm) -> {
-									WorkspaceEdit edit = getRenameEdit(oldPackageFragment, newPackageName, pm);
-									root[0] = ChangeUtil.mergeChanges(root[0], edit, true);
-								}, oldPackageFragment.getSchedulingRule(), IResource.NONE, renameMonitor);
-							}
-						}
-					} catch (CoreException e) {
-						JavaLanguageServerPlugin.logException("Failed to compute the package rename update", e);
-					} finally {
-						renameMonitor.done();
+			IPackageFragment oldPackageFragment = resolvePackageFragment(oldLocation, sourcePaths);
+			SubMonitor renameMonitor = submonitor.split(100);
+			try {
+				if (oldPackageFragment != null && !oldPackageFragment.isDefaultPackage() && oldPackageFragment.getResource() != null) {
+					String oldPackageName = oldPackageFragment.getElementName();
+					int lastDot = oldPackageName.lastIndexOf(".");
+					String newPackageName = lastDot < 0 ? newLocation.lastSegment() :
+						oldPackageName.subSequence(0, lastDot + 1) + newLocation.lastSegment();
+					oldPackageFragment.getResource().refreshLocal(IResource.DEPTH_INFINITE, null);
+					if (oldPackageFragment.exists()) {
+						ResourcesPlugin.getWorkspace().run((pm) -> {
+							WorkspaceEdit edit = getRenameEdit(oldPackageFragment, newPackageName, pm);
+							root[0] = ChangeUtil.mergeChanges(root[0], edit, true);
+						}, oldPackageFragment.getSchedulingRule(), IResource.NONE, renameMonitor);
 					}
-
-					break;
 				}
+			} catch (CoreException e) {
+				JavaLanguageServerPlugin.logException("Failed to compute the package rename update", e);
+			} finally {
+				renameMonitor.done();
 			}
 		}
 
 		submonitor.done();
 		return ChangeUtil.hasChanges(root[0]) ? root[0] : null;
 	}
 
+	private static WorkspaceEdit computeMoveEdit(FileRenameEvent[] moveEvents, SourcePath[] sourcePaths, IProgressMonitor monitor) {
+		IPath[] newPaths = Stream.of(moveEvents).map(event -> ResourceUtils.filePathFromURI(event.newUri)).toArray(IPath[]::new);
+		IPath destinationPath = getLongestCommonPath(newPaths);
+		if (destinationPath == null) {
+			return null;
+		}
+
+		IPackageFragment destinationPackage = resolvePackageFragment(destinationPath, sourcePaths);
+		if (destinationPackage == null) {
+			return null;
+		}
+
+		List<IJavaElement> javaElements = new ArrayList<>();
+		IPath[] oldPaths = Stream.of(moveEvents).map(event -> ResourceUtils.filePathFromURI(event.oldUri)).toArray(IPath[]::new);
```

Comment:
instead of using an intermediary IPath[] array, you can still manipulate the stream, filter the file paths, map to ICompilationUnits ...

Reply:
good suggestion.